ethcore/src/block.rs

```diff
@@ -278,10 +282,55 @@ impl<'x> OpenBlock<'x> {
 		let gas_floor_target = cmp::max(gas_range_target.0, engine.params().min_gas_limit);
 		let gas_ceil_target = cmp::max(gas_range_target.1, gas_floor_target);
 		engine.populate_from_parent(&mut r.block.header, parent, gas_floor_target, gas_ceil_target);
+		Self::push_last_hash(&mut r.block, last_hashes, engine, &parent.hash())?;
 		engine.on_new_block(&mut r.block);
 		Ok(r)
 	}
 
+
+	fn push_last_hash(block: &mut ExecutedBlock, last_hashes: Arc<LastHashes>, engine: &Engine, hash: &H256) -> Result<(), Error> {
+		if block.fields().header.number() == engine.params().eip210_transition {
+			let state = block.fields_mut().state;
+			state.init_code(&engine.params().eip210_contract_address, engine.params().eip210_contract_code.clone())?;
+		}
+		if block.fields().header.number() >= engine.params().eip210_transition {
+			let env_info = {
+				let header = block.fields().header;
+				EnvInfo {
+					number: header.number(),
+					author: header.author().clone(),
+					timestamp: header.timestamp(),
+					difficulty: header.difficulty().clone(),
+					last_hashes: last_hashes,
+					gas_used: U256::zero(),
+					gas_limit: engine.params().eip210_contract_gas,
+				}
+			};
+			let mut state = block.fields_mut().state;
+			let contract_address = engine.params().eip210_contract_address;
+			let params = ActionParams {
+				code_address: contract_address.clone(),
+				address: contract_address.clone(),
+				sender: SYSTEM_ADDRESS.clone(),
+				origin: SYSTEM_ADDRESS.clone(),
+				gas: engine.params().eip210_contract_gas,
+				gas_price: 0.into(),
+				value: ActionValue::Transfer(0.into()),
+				code: state.code(&contract_address)?,
+				code_hash: state.code_hash(&contract_address)?,
+				data: Some(hash.to_vec()),
+				call_type: CallType::Call,
+			};
+			let mut ex = Executive::new(&mut state, &env_info, engine);
+			let mut substate = Substate::new();
+			let mut output = [];
+			if let Err(e) = ex.call(params, &mut substate, BytesRef::Fixed(&mut output), &mut NoopTracer, &mut NoopVMTracer) {
```

Comment:
As far as I understand the EIP this is more of a call with preserved state changes. Finalization is not required.